src/Specular/Internal/Incremental.purs

```diff
@@ -0,0 +1,410 @@
+  runEffectFn2 Node.set_value node (Optional.some val)
+  _ <- runEffectFn2 PQ.add globalRecomputeQueue (toSomeNode node)
+  pure unit
+
+readEvent :: forall a. Event a -> Node a
+readEvent (Event x) = x
+
+-- * Observers and dependents
+
+addObserver :: forall a. EffectFn2 (Node a) (Observer a) Unit
+addObserver = mkEffectFn2 \node observer -> do
+  oldRefcount <- runEffectFn1 Node.refcount node
+  observers <- runEffectFn1 Node.get_observers node
+  runEffectFn2 MutableArray.push observers observer
+  runEffectFn2 handleRefcountChange node oldRefcount
+
+removeObserver :: forall a. EffectFn2 (Node a) (Observer a) Unit
+removeObserver = mkEffectFn2 \node observer -> do
+  oldRefcount <- runEffectFn1 Node.refcount node
+  observers <- runEffectFn1 Node.get_observers node
+  runEffectFn2 MutableArray.remove observers observer
+  runEffectFn2 handleRefcountChange node oldRefcount
+
+addDependent :: forall a. EffectFn2 (Node a) SomeNode Unit
+addDependent = mkEffectFn2 \node dependent -> do
+--  trace $ "addDependent " <> show (Node.name' node) <> " -> " <> show (Node.name' dependent)
+
+  oldRefcount <- runEffectFn1 Node.refcount node
+  dependents <- runEffectFn1 Node.get_dependents node
+  runEffectFn2 MutableArray.push dependents dependent
+  runEffectFn2 handleRefcountChange node oldRefcount
+
+removeDependent :: forall a. EffectFn2 (Node a) SomeNode Unit
+removeDependent = mkEffectFn2 \node dependent -> do
+  oldRefcount <- runEffectFn1 Node.refcount node
+  dependents <- runEffectFn1 Node.get_dependents node
+  runEffectFn2 MutableArray.remove dependents dependent
+  runEffectFn2 handleRefcountChange node oldRefcount
+
+handleRefcountChange :: forall a. EffectFn2 (Node a) Int Unit
+handleRefcountChange = mkEffectFn2 \node oldRefcount -> do
+  newRefcount <- runEffectFn1 Node.refcount node
+  if oldRefcount == 0 && newRefcount > 0 then
+    runEffectFn1 connect node
+  else if oldRefcount > 0 && newRefcount == 0 then
+    runEffectFn1 disconnect node
+  else
+    pure unit
+
+  -- Update globalTotalRefcount
+  oldTotalRefcount <- runEffectFn1 Ref.read globalTotalRefcount
+  runEffectFn2 Ref.write globalTotalRefcount (oldTotalRefcount - oldRefcount + newRefcount)
+
+-- Preconditions:
+-- - node does not have value computed
+-- - node does not have any dependents
+--
+-- Postconditions:
+-- - all dependencies are connected and have value computed
+-- - node has value computed
+-- - node has correct height
+connect :: forall a. EffectFn1 (Node a) Unit
+connect = mkEffectFn1 \node -> do
+--  trace $ "connect " <> show (Node.name' node)
+
+  source <- runEffectFn1 Node.get_source node
+  dependencies <- source.dependencies
+
+  runEffectFn2 Array.iterate dependencies $ mkEffectFn1 \dependency -> do
+    runEffectFn2 addDependent dependency (toSomeNode node)
+    dependencyHeight <- runEffectFn1 Node.get_height dependency
+    ourHeight <- runEffectFn1 Node.get_height node
+    if dependencyHeight + 1 > ourHeight then do
```

Comment:
This check: `dependencyHeight + 1 > ourHeight` could potentially be problematic with `adjustedHeight` ? Assuming that it is possible to have a dependency for which: `height <= ourHeight` but `adjustedHeight > ourHeight`

Reply:
This should work: if the dependency has `adjustedHeight > height`, then before stabilization gets to us, it will get to `dependencyHeight`, at which point height of the dependncy will get adjusted, and our `adjustedHeight` will be recomputed accordingly.
We could use `adjustedHeight` of the dependency, to save the additional height adjustment of the current node, but I believe this doesn't affect correctness.

Reply:
I was thinking we might be computing the subtree of `ourNode` based on stale data before  our height gets adjusted during stabilization.